mirgecom/profiling.py

```diff
@@ -0,0 +1,229 @@
+            r = self._get_kernel_stats(program, t.kwargs)
+            time = t.cl_event.profile.end - t.cl_event.profile.start
+
+            new = ProfileResult(time, r.flops, r.bytes_accessed, r.footprint_bytes)
+
+            self.profile_results.setdefault(program, []).append(new)
+
+        self.profile_events = []
+
+    def tabulate_profiling_data(self) -> pytools.Table:
+        """Return a :class:`pytools.Table` with the profiling results."""
+        self._finish_profile_events()
+
+        tbl = pytools.Table()
+
+        tbl.add_row(["Function", "Calls",
+            "Time_min [s]", "Time_avg [s]", "Time_max [s]",
+            "GFlops/s_min", "GFlops/s_avg", "GFlops/s_max",
+            "BWAcc_min [GByte/s]", "BWAcc_mean [GByte/s]", "BWAcc_max [GByte/s]",
+            "BWFoot_min [GByte/s]", "BWFoot_mean [GByte/s]", "BWFoot_max [GByte/s]"])
+
+        # Precision of results
+        g = ".4g"
+
+        from statistics import mean
+
+        for key, value in self.profile_results.items():
+            num_values = len(value)
+
+            times = [v.time / 1e9 for v in value]
+            flops = [v.flops / 1e9 for v in value]
+
+            flops_per_sec = [f / t for f, t in zip(flops, times)]
+
+            bytes_accessed = [v.bytes_accessed / 1e9 for v in value]
+            footprint_bytes = [v.footprint_bytes / 1e9 for v in value
+                if v.footprint_bytes is not None]
+
+            bandwidth_access = [b / t for b, t in zip(bytes_accessed, times)]
+
+            if len(footprint_bytes) > 0:
+                bandwidth_footprint = [b / t for b, t in zip(footprint_bytes, times)]
+                bandwidth_footprint_val = {"min": f"{min(bandwidth_footprint):{g}}",
+                    "mean": f"{mean(bandwidth_footprint):{g}}",
+                    "max": f"{max(bandwidth_footprint):{g}}"}
+            else:
+                bandwidth_footprint_val = {"min": "n/a", "mean": "n/a", "max": "n/a"}
+
+            tbl.add_row([key.name, num_values,
+                f"{min(times):{g}}", f"{mean(times):{g}}", f"{max(times):{g}}",
+                f"{min(flops_per_sec):{g}}", f"{mean(flops_per_sec):{g}}",
+                f"{max(flops_per_sec):{g}}",
+                f"{min(bandwidth_access):{g}}", f"{mean(bandwidth_access):{g}}",
+                f"{max(bandwidth_access):{g}}",
+                f"{bandwidth_footprint_val['min']}",
+                f"{bandwidth_footprint_val['mean']}",
+                f"{bandwidth_footprint_val['max']}"])
+
+        return tbl
+
+    def _get_kernel_stats(self, program, kwargs: dict) -> ProfileResult:
+        # We need a tuple to index the cache
+        args_tuple = tuple(
+            (key, value.shape) if hasattr(value, "shape") else (key, value)
+            for key, value in kwargs.items())
+
+        # Are kernel stats already in the cache?
+        if program in self.kernel_stats and args_tuple in self.kernel_stats[program]:
+            return self.kernel_stats[program][args_tuple]
```

Comment:
IMO the Python-idiomatic way would be like this:
```py
try:
    return self.kernel_stats[program][args_tuple]
except KeyError:
```
Then the fast (cache hit) path does not involve redundant lookups.

Reply:
Fixed.

Reply:
Good point, fixed.